### DISCUSSION TITLE
Annotations not working with constructor

### DISCUSSION THREAD

**Discussion by lacoonte** [nartc/mapper#363] [Q&A]
I have two classes:
```javascript
export default class Timetable {
  @AutoMap({ typeFn: () => Workday })
  workdays: Workday[];
  startDate: Date;
  endDate: Date;
  constructor(workdays: Workday[], startDate: Date, endDate: Date) {
    this.workdays = workdays;
    this.startDate = startDate;
    this.endDate = endDate;
  }
}
```
and:
```javascript
export default class Workday {
  @AutoMap({ typeFn: () => Number })
  day: WeekDay;
  @AutoMap()
  startTime: Date;
  @AutoMap()
  endTime: Date;
  constructor(
    day: WeekDay,
    startTime: Date,
    endTime: Date
  ) {
    this.day = day;
    this.startTime = startTime;
    this.endTime = endTime;
  }
}
```
Unless I remove the constructor in workday, i get the following error over "typeFn" in Timetable:
`Type '() => typeof Workday' is not assignable to type '() => Constructible<any>'.
  Type 'typeof Workday' is not assignable to type 'Constructible<any>'.
    Types of parameters 'day' and 'args' are incompatible.
      Type 'unknown' is not assignable to type 'number'.ts(2322)
automap.decorator.d.ts(7, 5): The expected type comes from property 'typeFn' which is declared here on type 'AutoMapOptions'`

WeekDay is an enum with number value. If I remove Workday's constructor, there's no error. Do you know why is this happening?
Thanks in advance

**Comment by micalevisk** [comment]
how's your mapping (`Mapper#createMap` calls)?

**Reply by lacoonte** [reply]
I don't have any. The error appears in Timetable.ts itself, over the third line.
`@AutoMap({ typeFn: () => Workday })`

**Reply by micalevisk** [reply]
I meant, the `mapper.createMap` lines

https://automapperts.netlify.app/docs/fundamentals/#mapping

**Reply by micalevisk** [reply]
btw how you're importing `Workday` from that `Timetable.ts` file?

**Reply by lacoonte** [reply]
I don't have any call yet, I'm just annotating classes. I'm importing it like:
`import Workday from "./Workday";`
The error appears in the Typescript analysis of the file:
![Screenshot from 2021-10-17 19-55-38](https://user-images.githubusercontent.com/21204482/137639148-8afb958a-23dc-411a-99e2-86a9c14be859.png)

**Reply by micalevisk** [reply]
I'm using `@AutoMap` with constructor just fine with TS 4.0.5, and `@automapper/*` v4 (the latest one is v6)

Which version of typescript are you using?

**Reply by micalevisk** [reply]
I've managed to reproduce your issue [in this TS playground](https://www.typescriptlang.org/play?ts=4.4.3&ssl=29&ssc=36&pln=29&pc=15#code/C4TwDgpgBAIglgY2HA9gOwIYCcQB4AqAfFALxQDeUA2gNYQhRxpR0goBmU+AugPwBcUAK5oaaFAHdmAXwDcAKHlNgELOwwJo+LBjQBndiiwBbVQFkIwDABMMVgGIbgRvPKhczKaxAA2UCAAeKmjWerCIyOjYePie3j7EZLog8sTkblAA+pkAcgCCAEr4AMKZeQCq+ADyZnkACnUAogWZZo34eTB5HZn2ecXVBQCaAlAAFACUpMTwSKiYOARxvoQK0orKquqaUMXoesBYQnMARj4QS16+-kEQIWGzkQsxywmkUMmEGYHBoVw6+kMJnMlhsdgwjiQLku8TSGTQEAk4wAdKjsABzPSCERiSRoKjcCaCWJXHxrDZoFRqDTQPJCZxmDBgKpgJ5hdLuUCQexoUaTaa7faHY7IM4QcnydgiOboKB0hlMsYoVl6UbylCM5ms+Z6IlQOpYZWqUAwCAIIx2IwUKBYSxCLDMfkkNLSKDrRTyBA+DB6ML4OCmKxiigZAAC6s1Y0oXIgPMETuIAHUjDRbAxpBMMhIU2msVBk1hUxgQASFBkDthgDA7BBBNWVAp3N8QvXa7Aa42oOb9MKoVgxtnC7nBAWiyXuAAaKAVrBVmt1mtTu7WVsLlRTDnuKDAAAWcD0yMHY7CZCPuc7nL3B5nc5U7xvrYv26vyOXrfeb47GXW7q9PrCo5piG7jhvSGqKpm7g3v6phruKGRhhGEHNtYMFto+ijuN2BxHH2YwZFBVizmhcETgR-ghCR7YqBkG7kbu+7ItBAbQGQzGmE+DEHsuaEfpRLGdj+QA) but didn't find any wayt to fix it

**Reply by nartc** [reply]
This is probably a bug with `Constructible` type. I've never had a constructor in my models that are subject to mapping.

**Reply by micalevisk** [reply]
you're right. After upgrading `@automapper/*` to 6.2.1 from 4.2.1 I start getting that same error.

**Reply by micalevisk** [reply]
actually the error I was getting isn't related with `Constructible` type, they come from breaking changes; my bad.

I'm not sure how I was able to using this pattern just fine in both versions, tbh:

![image](https://user-images.githubusercontent.com/13461315/137657280-be698392-ab04-4061-bca8-d3ee36e1b3dc.png)

![image](https://user-images.githubusercontent.com/13461315/137657380-88777d84-5423-4273-9d6f-0fc0b976a45b.png)

**Reply by nartc** [reply]
Maybe your tsconfig isn't as strict

**Reply by micalevisk** [reply]
you're right! I'm using `strictFunctionTypes: false` + `strict: false` :+1:

**Reply by lacoonte** [reply]
I don't know that much about TypeScript (I started using it recently), but wouldn't it be possible to constraint the constructor arguments in Constructible to the ones of the class' constructor? (Maybe using reflection)

**Reply by lacoonte** [reply]
I found this, don't know if it is possible to use it https://www.typescriptlang.org/docs/handbook/utility-types.html#constructorparameterstype

**Reply by nartc** [reply]
@alvarorece it is possible but it's not necessary for the case of `@AutoMap` and `Constructible`. I've already created a bug to address this so it should be fixed in the next release.
